ml-agents-envs/mlagents_envs/registry/binary_utils.py

```diff
@@ -0,0 +1,214 @@
+    platform dependent : It will only return a Unity executable compatible with the
+    computer's OS. If no executable is found, None will be returned.
+    :param name: The name/identifier of the executable
+    :param url: The url the executable was downloaded from (for verification)
+    """
+    _, bin_dir = get_tmp_dir()
+    extension = None
+
+    if platform == "linux" or platform == "linux2":
+        extension = "*.x86_64"
+    if platform == "darwin":
+        extension = "*.app"
+    if platform == "win32":
+        extension = "*.exe"
+    if extension is None:
+        raise NotImplementedError("No extensions found for this platform.")
+    url_hash = "-" + hashlib.md5(url.encode()).hexdigest()
+    path = os.path.join(bin_dir, name + url_hash, "**", extension)
+    candidates = glob.glob(path, recursive=True)
+    if len(candidates) == 0:
+        return None
+    else:
+        for c in candidates:
+            # Unity sometimes produces another .exe file that we must filter out
+            if "UnityCrashHandler64" not in c:
+                return c
+        return None
+
+
+def get_tmp_dir() -> Tuple[str, str]:
+    """
+    Returns the path to the folder containing the downloaded zip files and the extracted
+    binaries. If these folders do not exist, they will be created.
+    :retrun: Tuple containing path to : (zip folder, extracted files folder)
+    """
+    MLAGENTS = "ml-agents-binaries"
+    TMP_FOLDER_NAME = "tmp"
+    BINARY_FOLDER_NAME = "binaries"
+    mla_directory = os.path.join(tempfile.gettempdir(), MLAGENTS)
+    if not os.path.exists(mla_directory):
+        os.makedirs(mla_directory)
+        os.chmod(mla_directory, 16877)
+    zip_directory = os.path.join(tempfile.gettempdir(), MLAGENTS, TMP_FOLDER_NAME)
+    if not os.path.exists(zip_directory):
+        os.makedirs(zip_directory)
+        os.chmod(zip_directory, 16877)
+    bin_directory = os.path.join(tempfile.gettempdir(), MLAGENTS, BINARY_FOLDER_NAME)
+    if not os.path.exists(bin_directory):
+        os.makedirs(bin_directory)
+        os.chmod(bin_directory, 16877)
+    return (zip_directory, bin_directory)
+
+
+def download_and_extract_zip(url: str, name: str) -> None:
+    """
+    Downloads a zip file under a URL, extracts its contents into a folder with the name
+    argument and gives chmod 755 to all the files it contains. Files are downloaded and
+    extracted into special folders in the temp folder of the machine.
+    :param url: The URL of the zip file
+    :param name: The name that will be given to the folder containing the extracted data
+    """
+    zip_dir, bin_dir = get_tmp_dir()
+    url_hash = "-" + hashlib.md5(url.encode()).hexdigest()
+    binary_path = os.path.join(bin_dir, name + url_hash)
+    if os.path.exists(binary_path):
+        shutil.rmtree(binary_path)
+
+    # Download zip
+    try:
+        request = urllib.request.urlopen(url)
```

Comment:
Please pass a timeout here, otherwise it could hang indefinitely. I'd say 30 seconds as a default, with optional parameters to override?